### PR TITLE
Feature/rpc networking

### PR DESCRIPTION
**Pull request by DustedHam**

### CODE REVIEW

**Comment by DustedHam** [comment]
The build isn't actually failing on anything important, so that can be ignored.
I've got to figure out why the build steps are broken is all.

File: Convert.cs [hunk 1]
```diff
@@ -0,0 +1,165 @@
+﻿using System;
+
+namespace Saltbox.Net
+{
+    public class Convert
```
Comment (DustedHam) [inline Convert.cs]: This should be static

File: RpcClient.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        public async void Connect(string host, int port)
+        {
+            TcpClient client = new TcpClient();
+            await client.ConnectAsync(host, port);
+
+            if (!client.Connected)
+            {
+                OnConnectFail?.Invoke();
```
Comment (JohnGames) [inline RpcClient.cs]: The reason would be appreciated.

File: RpcClient.cs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            }
+
+            NetworkStream stream = client.GetStream();
+
+            AsyncQueue<IJob> sendQueue = new AsyncQueue<IJob>();
+            TRemote encoder = RpcEventEncoder<TRemote>.Create(new Sender(stream));
+            TRemote remoteQueuer = AsyncJobQueuer<TRemote>.Create(encoder, sendQueue);
+
+            _ = Task.Factory.StartNew(() =>
```
Comment (JohnGames) [inline RpcClient.cs]: Consider adding a comment stating going off Unity Main Thread.

Comment (DustedHam) [inline RpcClient.cs]: this should be set to long running

File: RpcClient.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            {
+                Task.WaitAll(Send(client, sendQueue), Receive(stream));
+            });
+
+            OnConnected?.Invoke(remoteQueuer);
+        }
+
+        public IEnumerator ExecuteLocalRoutine()
+        {
+            Task<IJob> awaiting;
```
Comment (JohnGames) [inline RpcClient.cs]: Please adjust from yield return null to a yield return new wait until

File: RpcClient.cs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                    read += await stream.ReadAsync(buffer, read, length - read);
+                }
+                while (read < length);
+                decoder.Invoke(id, buffer);
+            }
+
+            OnDisconnected.Invoke();
+        }
+
+        private class Sender : IRpc
```
Comment (JohnGames) [inline RpcClient.cs]: I personally would prefer this as a partial in a separate file.

Comment (DustedHam) [reply RpcClient.cs]: I'm going to leave this as is, I don't think it's big enough to warrant a partial class. And I typically avoid partials.

File: Convert.cs [hunk 6]
```diff
@@ -0,0 +1,170 @@
+﻿using System;
+
+namespace Saltbox.Net
+{
+    public static class Convert
```
[the remark this answers is not in the document]
Comment (DustedHam) [reply Convert.cs]: you're not wrong, I'll look into setting up proper tests and build steps. but that will have to wait for another PR I think.

File: Convert.cs [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+            from(data, ref position, out val);
+            value = *(float*)&val;
+        }
+
+        public static unsafe void from(byte[] data, ref int position, out long value)
+        {
+            fixed (byte* pbyte = &data[position])
+            {
+                int i1 = (*pbyte << 24) | (*(pbyte + 1) << 16) | (*(pbyte + 2) << 8) | (*(pbyte + 3));
+                int i2 = (*(pbyte + 4) << 24) | (*(pbyte + 5) << 16) | (*(pbyte + 6) << 8) | (*(pbyte + 7));
```
[the remark this answers is not in the document]
Comment (DustedHam) [reply Convert.cs]: where these are all bit-shift operations I find the literals easier to follow, so I'm going to leave these as is, unless there's a major issue.

File: Generation/GenerationData.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
+﻿using System.Collections.Generic;
+using System.Reflection;
+
+using Saltbox.Collections;
+
+namespace Saltbox.Net.Generation
+{
+    internal class GenerationData
+    {
+        private readonly List<string> assembies = new List<string>();
```
Comment (mcdon00) [inline Generation/GenerationData.cs]: assembies -> assemblies

**Pull request merged**